securebanking-openbanking-uk-gateway-conversion-filter/src/main/java/com/forgerock/securebanking/uk/gateway/conversion/filter/IntentConverterFilter.java

```diff
@@ -0,0 +1,287 @@
+        logger.error("Conversion to OB Object filter Error\n", e);
+        response.setCause(e);
+        response.setEntity(e.getMessage().getBytes());
+        response.setStatus(Status.BAD_REQUEST);
+    }
+
+    private void process(Request request, Response response, String entity) throws Exception {
+        String jsonPayload = entity;
+        if (Strings.isNullOrEmpty(jsonPayload)) {
+            throw new Exception("The entity to be converted should not be null");
+        }
+        OBVersion obVersion = ApiVersionUtils.getOBVersion(request.getUri().asURI());
+        Object objectMapped = convert(getIntentType(jsonPayload), jsonPayload, obVersion);
+        logger.info("Result object {}", objectMapped.getClass().getSimpleName());
+        logger.debug("objectMapped {}", objectMapped);
+        if (resultTo.contains(MessageType.REQUEST)) {
+            setEntity(request, toBytes(objectMapped));
+        }
+        if (resultTo.contains(MessageType.RESPONSE)) {
+            setEntity(response, toBytes(objectMapped));
+        }
+    }
+
+    private void setEntity(final Message<?> message, byte[] content) {
+        message.setEntity(content);
+        if (!message.getHeaders().containsKey(ContentTypeHeader.NAME)) {
+            message.getHeaders().put(ContentTypeHeader.NAME, APPLICATION_JSON_CHARSET_UTF_8);
+        }
+    }
+
+    private IntentType getIntentType(String payload) throws Exception {
+
+        Map<String, Object> map = GenericConverterMapper.getMapper().readValue(payload, HashMap.class);
+        if (map.get("Data") == null) {
+            throw new Exception("The entity doesn't have 'Data' Object to identify the intent type");
+        }
+        String consentId = ((Map<String, String>) map.get("Data")).get("ConsentId");
+        if (consentId == null) {
+            throw new Exception("The entity doesn't have 'ConsentId' to identify the intent type");
+        }
+        IntentType intentType = IntentType.identify(consentId);
+        if (intentType == null) {
+            throw new Exception("It cannot be possible to identify the intent type with the consentId " + consentId);
+        }
+        return intentType;
+    }
+
+    private String getEntity(final Message<?> message, final Bindings bindings) throws Exception {
+        if (entity != null) {
+            logger.debug("Payload from optional entity {}", entity);
+            AtomicReference<String> content = new AtomicReference<>();
+            entity.evalAsync(bindings)
+                    .then(result -> {
+                        if (result == null) {
+                            throw new Exception("Unexpected null result from this expression " + entity);
+                        }
+                        return result;
+                    }, neverThrown()).thenOnResult(result -> {
+                        content.set(result);
+                    });
+            return content.get();
```

Comment:
this will only work if the evalAsync determines that it can evaluate the expression without an async call i.e. behaves the same as calling the eval synchronous function.

if evalAsync has to do an async call then content.get() will mostly likely return null (as the async callback thenOnResult won't have executed to call content.set yet).

So you need to determine if the eval ever needs to do an async call, if it does not then you can swap evalAsync for eval which returns a value. Otherwise you need to restructure the code to do the relevant processing in the promise callbacks / chain more callbacks.

The eval method logs an error if it thinks you should be using the async version, so this can help with investigation:
```
 public T eval(final Bindings bindings) {
        if (valueExpression.isDeferred()) {
            logger.error("Trying to invoke synchronously an expression ('{}') that requires async", original);
        }
        return evaluateNow(bindings);
    }
```

Reply:
Yes, we no need the async method. change done.